### PR TITLE
Add cluster setting to enable failure store

### PR DESCRIPTION
**Pull request by PeteGillinElastic**
This setting enables or disables the failure store for data streams based on matching the data stream name against a list of patterns. It acts as a default, and is overridden if the failure store is explicitly enabled or disabled either in a component template or using the data stream options API.

(See the PR for explanations of some of the changes here.)

### CODE REVIEW

File: modules/data-streams/src/yamlRestTest/resources/rest-api-spec/test/data_stream/200_require_data_stream.yml
Comment (PeteGillinElastic) [inline modules/data-streams/src/yamlRestTest/resources/rest-api-spec/test/data_stream/200_require_data_stream.yml:1]: Renaming this as a drive-by because there were two `190_*` files before.

File: modules/data-streams/src/yamlRestTest/resources/rest-api-spec/test/data_stream/210_rollover_failure_store.yml
Comment (PeteGillinElastic) [inline modules/data-streams/src/yamlRestTest/resources/rest-api-spec/test/data_stream/210_rollover_failure_store.yml:1]: Renaming this as a drive-by because there were two `190_*` files before.

File: server/src/main/java/org/elasticsearch/cluster/metadata/DataStreamFailureStoreSettings.java
Comment (gmarouli) [inline server/src/main/java/org/elasticsearch/cluster/metadata/DataStreamFailureStoreSettings.java:65]: I was a bit surprised to see the `isInternal` flag here. It feels to me like we are leaking information. I can walk you through how I was separating the concerns in my head and you can tell me if you agree or not.

The `DataStreamFailureStoreSettings` is only holding the failure store as default cluster setting, and since that's a list of regexes I think the predicate very well captures this. I think here, we shouldn't bother with the internal data streams.

The logic that this setting doesn't apply to internal data streams, I think, belongs with the resolution of the effective failure store enabled flag. There we pose the questions:
- "is the flag explicitly set?", if not,
- "is the data stream internal?", if not,
then we check with the settings. Does this make sense?

File: server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java
Comment (gmarouli) [inline server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java:476]: I would recommend check the `isInternal` flag here. The next comment explains a bit better the reasons behind my thinking.

**Review by gmarouli — COMMENTED** [review]
Great work @PeteGillinElastic ! It really good, I added some minor comments and opened one discussion about when to check `isInternal`. Curious to hear your thoughts.

File: modules/data-streams/src/javaRestTest/java/org/elasticsearch/datastreams/DataStreamOptionsIT.java
Comment (PeteGillinElastic) [reply modules/data-streams/src/javaRestTest/java/org/elasticsearch/datastreams/DataStreamOptionsIT.java:239]: That's a good point about the `disable` method being misleading. Looking at it, maybe the simplest thing to do is to have a single `setDataStreamFailureStoreClusterSetting(String)` method. That way, the call sites are still one-liners, and it's hopefully pretty obvious what we're doing in each case.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/datastreams/DataStreamFeatureSetUsage.java
Comment (PeteGillinElastic) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/datastreams/DataStreamFeatureSetUsage.java:56]: I've added this to my TODO list. But I don't know anything about it, so if I'm going to do it then I'll need somebody to show me where it lives!

File: server/src/test/java/org/elasticsearch/cluster/metadata/DataStreamTests.java
Comment (PeteGillinElastic) [reply server/src/test/java/org/elasticsearch/cluster/metadata/DataStreamTests.java:2301]: Good call, I'll make this change. I'm going for `regular` because I found `user` a little confusing (what user? what's this got to do with them?).

File: server/src/test/java/org/elasticsearch/cluster/metadata/DataStreamTests.java
Comment (PeteGillinElastic) [reply server/src/test/java/org/elasticsearch/cluster/metadata/DataStreamTests.java:2320]: Yeah, I'd initially tried to do that, and run into the problem that `SystemDataStreamDescriptor` won't like you set a non-dot-prefixed name, making it impossible to test the system data stream logic independently of the dot-prefix logic — the method would return false for `.some-system-data-stream` whether or not the `SystemIndices` object reckoned it was a system data stream or not.

By that reasoning, we could actually dispense with the `SystemIndices` check in the code under test, but it seems like good defensive programming to do it anyway.

But it occurs to me that I can mock `SystemIndices` and get it to behave in the way that the real one never will. See what you think.

File: server/src/test/java/org/elasticsearch/cluster/metadata/DataStreamTests.java
Comment (PeteGillinElastic) [reply server/src/test/java/org/elasticsearch/cluster/metadata/DataStreamTests.java:2265]: Ah, go on, why not!

File: server/src/main/java/org/elasticsearch/cluster/metadata/DataStreamFailureStoreSettings.java
Comment (PeteGillinElastic) [reply server/src/main/java/org/elasticsearch/cluster/metadata/DataStreamFailureStoreSettings.java:65]: Yeah, I totally see what you mean.

The reason I did it this way is that it would be a mistake to ever call `settings. failureStoreEnabledForDataStreamName(name)` without also doing the `isInternal` check. By putting the check in this method, we force the caller to think about it. Without, it would be possible for the caller to forget about that part of the requirements.

*However*, when I originally went through that thought process, I thought that we were going to be calling this method in multiple places. I've actually managed to common up the methods in `DataStream` so that it's only called from one place. So now my argument doesn't really hold water, as we have exactly one place we have to remember it either way 🙂.

So I think I'll take your advice and change it to the more intuitive thing.

File: server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java
Comment (PeteGillinElastic) [reply server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java:476]: Done, see discussion on the other comment.

**Comment by PeteGillinElastic** [comment]
Thanks for the comments @gmarouli . I think I've addressed everything.

File: modules/data-streams/src/javaRestTest/java/org/elasticsearch/datastreams/DataStreamOptionsIT.java
Comment (gmarouli) [reply modules/data-streams/src/javaRestTest/java/org/elasticsearch/datastreams/DataStreamOptionsIT.java]: Sounds like a plan!

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/datastreams/DataStreamFeatureSetUsage.java
Comment (gmarouli) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/datastreams/DataStreamFeatureSetUsage.java:56]: Happy to show you around when time comes!

File: server/src/test/java/org/elasticsearch/cluster/metadata/DataStreamTests.java
Comment (gmarouli) [reply server/src/test/java/org/elasticsearch/cluster/metadata/DataStreamTests.java]: Regular is fine too, just for context, the idea behind the prefix `user` is that these cluster settings apply to user data, meaning data streams containing data that belong to the user and not internal data that elasticsearch needs to function.

File: server/src/test/java/org/elasticsearch/cluster/metadata/DataStreamTests.java
Comment (gmarouli) [reply server/src/test/java/org/elasticsearch/cluster/metadata/DataStreamTests.java]: > By that reasoning, we could actually dispense with the SystemIndices check in the code under test, but it seems like good defensive programming to do it anyway. 

That is probably there because we might have legacy indices, but data streams are pretty new so I don't think we will encounter this. That being said, I prefer to have it for completeness for now.

I like the idea of mocking if it is not too much trouble.

File: modules/data-streams/src/yamlRestTest/resources/rest-api-spec/test/data_stream/220_failure_store_cluster_setting.yml
Comment (gmarouli) [inline modules/data-streams/src/yamlRestTest/resources/rest-api-spec/test/data_stream/220_failure_store_cluster_setting.yml:27]: <3

File: modules/data-streams/src/yamlRestTest/resources/rest-api-spec/test/data_stream/220_failure_store_cluster_setting.yml
Comment (gmarouli) [inline modules/data-streams/src/yamlRestTest/resources/rest-api-spec/test/data_stream/220_failure_store_cluster_setting.yml:102]: Does it not accept `null` here? It might be better to completely reset it.

File: modules/data-streams/src/yamlRestTest/resources/rest-api-spec/test/data_stream/220_failure_store_cluster_setting.yml
Comment (gmarouli) [inline modules/data-streams/src/yamlRestTest/resources/rest-api-spec/test/data_stream/220_failure_store_cluster_setting.yml:127]: Nit: Maybe here we can just create the data stream with `indices.create_data_stream`.

**Review by gmarouli — COMMENTED** [review]
Awesome work! For me it's almost ready, I left some nits and opened a short, I hope, discussion about the yaml test node feature.

File: modules/data-streams/src/javaRestTest/java/org/elasticsearch/datastreams/DataStreamOptionsIT.java
Comment (PeteGillinElastic) [reply modules/data-streams/src/javaRestTest/java/org/elasticsearch/datastreams/DataStreamOptionsIT.java:158]: > Nit: should we set setDataStreamFailureStoreClusterSetting("does-not-match-failure-data-stream"); a few lines earlier and perform the checks only once?

Right now, we're testing all combinations of [enabled in options, disabled in options, not specified in options] ⊗ [enabled by settings, not enabled by settings]. I guess you're saying that we could reduce that by not bothering to check enabled in options and enabled by settings when we expect the same outcome with enabled in settings and not enabled by settings and the latter is the 'stronger' test? I see what you mean — but, on the other hand, why _not_ test all the combinations? Surely a fraction of a second on the test execution time isn't a problem?

> The difference between having the cluster setting set to a non-matching value or null shouldn't influence this test, so we demonstrate that by this randomisation.

I'd actually rather test both a non-matching value and a null value, rather than randomly testing one of the two. As discussed in a Slack thread a while ago, I think we over-used randomization in tests a bit, and if we care about two behaviours then we should just always test both of them. According to my highly unscientific benchmarking, adding tests for the null setting with all three options values costs us about 75ms, which I think is okay. I'll push a commit with that now.

File: modules/data-streams/src/javaRestTest/java/org/elasticsearch/datastreams/DataStreamOptionsIT.java
Comment (PeteGillinElastic) [reply modules/data-streams/src/javaRestTest/java/org/elasticsearch/datastreams/DataStreamOptionsIT.java:182]: Same thought process as above... Let me know if I've misunderstood or if you disagree!

File: modules/data-streams/src/main/java/org/elasticsearch/datastreams/action/TransportGetDataStreamsAction.java
Comment (PeteGillinElastic) [reply modules/data-streams/src/main/java/org/elasticsearch/datastreams/action/TransportGetDataStreamsAction.java:172]: Yeah, I was a bit torn about the right place to do that check. I do have an `assert` in `DataStreamFailureStoreSetting.failureStoreEnabledForDataStreamName` to ensure that all calls are guarded by a feature flag check. I wondered whether we could move the check itself further up the stack.

The thing is that I don't think we can put this check into `DataStream.isFailureStoreExplicitlyEnabled`, because some of the callers don't want it, and I thought it would be confusing if `DataStream.isFailureStoreEffectivelyEnabled` did the check but `DataStream.isFailureStoreExplicitlyEnabled` didn't.

Also, I don't think we should sweat this decision too much, since wherever we put the check, we'll be taking it out before too long (🤞) when we remove the feature flag.

Let me know if you feel strongly that we should change this.

File: modules/data-streams/src/test/java/org/elasticsearch/datastreams/action/TransportGetDataStreamsActionTests.java
Comment (PeteGillinElastic) [reply modules/data-streams/src/test/java/org/elasticsearch/datastreams/action/TransportGetDataStreamsActionTests.java:409]: Yeah, the Hamcrest `contains` matcher is actually really confusing: This looks like it's checking that at least one element in the list is `false`, but it's actually checking that the list contains exactly one element which is `false`. (I kind of hate Hamcrest, by the way. More carefully designed libraries like AssertJ and Truth use `containsExactly` for this assertion.)

So, while I think what I've written is doing logically the right thing, I agree that it's not readable code. My excuse is that I was following the pattern from elsewhere in this class (see e.g. the assertion in `testGetTimeSeriesDataStream`). But there's no reason not to improve things! I'll push something more explicit.

Comment (gmarouli) [reply modules/data-streams/src/test/java/org/elasticsearch/datastreams/action/TransportGetDataStreamsActionTests.java]: Thank you, I think there is something similar to `containsExactly`. I can look it up if you want.

File: modules/data-streams/src/javaRestTest/java/org/elasticsearch/datastreams/DataStreamOptionsIT.java
Comment (gmarouli) [reply modules/data-streams/src/javaRestTest/java/org/elasticsearch/datastreams/DataStreamOptionsIT.java:162]: Ok, no strong feelings about this.

File: modules/data-streams/src/yamlRestTest/resources/rest-api-spec/test/data_stream/220_failure_store_cluster_setting.yml
Comment (PeteGillinElastic) [reply modules/data-streams/src/yamlRestTest/resources/rest-api-spec/test/data_stream/220_failure_store_cluster_setting.yml]: So, yeah, you're right, I was being a bit lazy and piggy-backing on the `failure_store_in_template` capability. You only added that last week, so I do think they have a good chance of appearing in the same version... But it's not certain. I've pushed a commit which adds a node feature for this... Let me know what you think.

I do have a couple of follow-up questions:

1. Should do `assumeTrue(clusterHasFeature(...))` in the Java REST test `DataStreamOptionsIT`? It would be easy enough, but we don't seem to have an equivalent check for the API capability right now, so I'm not sure.
2. I'm not quite sure _why_ we have this `requires` block in the YAML test. Do we ever run tests from `HEAD` against an older server or a server without all feature flags enabled? From my limited experimentation, it seems like the effect of the `requires` block would be to make the test fail-fast if we did, anyway.

File: modules/data-streams/src/yamlRestTest/resources/rest-api-spec/test/data_stream/220_failure_store_cluster_setting.yml
Comment (PeteGillinElastic) [reply modules/data-streams/src/yamlRestTest/resources/rest-api-spec/test/data_stream/220_failure_store_cluster_setting.yml]: Oh, yeah, that is neater — thanks.

File: modules/data-streams/src/yamlRestTest/resources/rest-api-spec/test/data_stream/220_failure_store_cluster_setting.yml
Comment (PeteGillinElastic) [reply modules/data-streams/src/yamlRestTest/resources/rest-api-spec/test/data_stream/220_failure_store_cluster_setting.yml]: Definitely simplifies things and makes it more clear what we're actually testing. Thanks. (I'd looked to see if this action existed, but I think I'd been looking for something like `put_data_stream` rather than `create_` so I didn't find it, and figured I'd do it the long way around... I should have looked harder!)

**Comment by PeteGillinElastic** [comment]
Thanks again @gmarouli , all comments responded to now.

File: modules/data-streams/src/yamlRestTest/resources/rest-api-spec/test/data_stream/220_failure_store_cluster_setting.yml
Comment (gmarouli) [reply modules/data-streams/src/yamlRestTest/resources/rest-api-spec/test/data_stream/220_failure_store_cluster_setting.yml]: One of the differences between YAML tests and Java REST tests is that YAML tests are also run in bwc, meaning they run against older versions of elasticsearch. This is why we use these "requirements" to ensure that the tests are going to be run against versions that support the feature.

Also this is why, you do not need this in Java REST test.

Now going back to the node feature. I think we need to think as well, do we need this node feature in the production code? If not, we can rethink if it's useful to add it or not.

Comment (PeteGillinElastic) [reply modules/data-streams/src/yamlRestTest/resources/rest-api-spec/test/data_stream/220_failure_store_cluster_setting.yml]: But BWC is about running an old test suite against a new cluster, right?

Feature flags are enabled (and can't be disabled) in all snapshot builds: https://github.com/elastic/elasticsearch/blob/8134c79ce8a11b26ee26d88ad7975c91a127af98/server/src/main/java/org/elasticsearch/common/util/FeatureFlag.java#L72. So long as you're running the tests with a snapshot build rather than a release build, they should all be on. Unless I've misunderstood... These tests are certainly running locally for me!

Comment (PeteGillinElastic) [reply modules/data-streams/src/yamlRestTest/resources/rest-api-spec/test/data_stream/220_failure_store_cluster_setting.yml]: > Now going back to the node feature. I think we need to think as well, do we need this node feature in the production code? If not, we can rethink if it's useful to add it or not.

I don't really understand the node features framework well enough to answer this question. I'll defer to you, or we can go ask someone who knows this stuff better.

Comment (PeteGillinElastic) [reply modules/data-streams/src/yamlRestTest/resources/rest-api-spec/test/data_stream/220_failure_store_cluster_setting.yml]: (I can easily revert that commit if we don't need it.)

Comment (gmarouli) [reply modules/data-streams/src/yamlRestTest/resources/rest-api-spec/test/data_stream/220_failure_store_cluster_setting.yml]: > But BWC is about running an old test suite against a new cluster, right?

No, if I understand it correctly, we run the new test suite to previous versions, including mixed clusters.

> Feature flags are enabled (and can't be disabled) in all snapshot builds:

That is true, but we also run the test suites as release, so we explicitly enable the feature flags because we might end up with mixed versions.

> I don't really understand the node features framework well enough to answer this question. I'll defer to you, or we can go ask someone who knows this stuff better.

There are certain features that need to be supported by all nodes before they can take effect. For example, hypothetically, we might want to allow users to use the failure store, only when all nodes support the failure store. Otherwise, you might have a node adding failure store but the bulk request goes to a different node that doesn't know about the failure store. This can be ensure via a node feature. 

A node feature can be used also in yaml tests to see if the feature is supported.

So, I was wondering here, considering that:

- We are still behind a feature flag.
- I think, without being 100% sure, that the capabilities of these APIs will remain there and will be in the same version as this.

Maybe it depends if we are planning to introduce a general node feature for the failure store. @jbaiera & @dakrone . Any thoughts? 

Could we bundle this under a failure store node feature when we remove the feature flag?

File: server/src/main/java/org/elasticsearch/cluster/metadata/DataStreamFailureStoreSettings.java
Comment (gmarouli) [reply server/src/main/java/org/elasticsearch/cluster/metadata/DataStreamFailureStoreSettings.java:71]: Ah! Such a good point that I completely missed. Apologies. Actually, we should only include this setting in the `ClusterSettings` if the feature flag is enabled. This will have as an effect that this setting won't exist in the system and if someone defines it will throw an error no matter what.

You can see commit https://github.com/elastic/elasticsearch/commit/01ed7de99ff22f92f08db11b666be88794f33829#diff-8e9e5bbba5094b994150b78efa335859b3435167c688e5e21d079dc92c198e6d, for reference.

**Comment by PeteGillinElastic** [comment]
I've just pushed a commit to do date math resolution on the name before matching it against the patterns in the cluster setting, as discussed on Slack. As part of this, I removed the test in `TransportBulkActionTests` which was directly calling `resolveFailureStoreFromTemplate`, since I don't think it added any more coverage than we already have from the test which calls `resolveFailureInternal`, and it's better to test via something closer to the public API. Let me know if you disagree with that.

File: server/src/main/java/org/elasticsearch/cluster/metadata/DataStreamFailureStoreSettings.java
Comment (PeteGillinElastic) [reply server/src/main/java/org/elasticsearch/cluster/metadata/DataStreamFailureStoreSettings.java:73]: Good call. I've pushed a commit which only registers the setting in `ClusterSettings` if the flag is enabled, and also changes `DataStreamFailureStoreSettings.create` to only `initializeAndWatch` the setting on the same condition as well (otherwise that call would fail).

**Comment by PeteGillinElastic** [comment]
Just rounding up where I think we are on this:

There is a thread on BWC, mixed cluster support, node features, and the like here: https://github.com/elastic/elasticsearch/pull/118662#discussion_r1888137526.

That's the only open issue that I'm aware of.

But we haven't heard from @jbaiera yet...

File: modules/data-streams/src/yamlRestTest/resources/rest-api-spec/test/data_stream/220_failure_store_cluster_setting.yml
Comment (PeteGillinElastic) [reply modules/data-streams/src/yamlRestTest/resources/rest-api-spec/test/data_stream/220_failure_store_cluster_setting.yml]: Yes, having slept on it, I think this:

> Could we bundle this under a failure store node feature when we remove the feature flag?

is the conclusion I'm leaning towards as well.

Reasoning:

1. We will not remove the feature flag until the implementation is complete.
2. Features are not, as I understand it, considered supported while they're behind a feature flag (the user would have to set the property or use a snapshot build, and figure out how to use it without documentation).
3. _Therefore_ we don't need to support the case of a mixed cluster in which the feature is enabled on some/all nodes and some/all nodes only have a partial implementation.
4. _Therefore_ there is no need for a node feature for the cluster setting bit itself, and I should revert the commit which added it.
5. _However_ we might want to support the case of a mixed cluster in which the feature is fully-implemented and enabled by default on some nodes and disabled on others (e.g. if we finish the implementation and remove the feature flag in 8.18, and the cluster has a mix of 8.17 and 8.18).
6. _Therefore_ we might need to use the node feature mechanism for the whole failure store feature.
7. _But_ that shouldn't be part of this PR — perhaps it should be part of the PR which removes the feature flag, just replacing checks on the feature flag with checks on the cluster feature (so it'll be enabled only if every node has the feature).

Comment (PeteGillinElastic) [reply modules/data-streams/src/yamlRestTest/resources/rest-api-spec/test/data_stream/220_failure_store_cluster_setting.yml:11]: As discussed in the meeting, I've removed the node feature and added a capability on the `PUT /_cluster/settings` API instead (with a comment mentioning that we plan to replace this when we remove the feature gate).

**Comment by PeteGillinElastic** [comment]
> Just rounding up where I think we are on this:
> 
> There is a thread on BWC, mixed cluster support, node features, and the like here: [#118662 (comment)](https://github.com/elastic/elasticsearch/pull/118662#discussion_r1888137526).
> 
> That's the only open issue that I'm aware of.
> 
> But we haven't heard from @jbaiera yet...

Okay, I've pushed changes for the above, so I think I've resolved all the issues I was aware of and this PR is ready for review again. 😮‍💨

**Review by jbaiera — APPROVED** [review]
I think everything LGTM pending final approval from @gmarouli. Sorry for jumping in on this late but great work on the discussions that have already happened so far!

**Review by gmarouli — APPROVED** [review]
LGTM, great work and great collaboration @PeteGillinElastic 🚀

**Comment by PeteGillinElastic** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.x|[](https://github.com/elastic/elasticsearch/pull/119132)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

**Pull request merged**
